File: leetcode/array/169.py
from typing import List


class Solution:

    def majorityElement(self, nums: List[int]) -> int:
        '''查找数组中的多数元素,多数元素:出现次数>n/2的元素.
        
        @Note:
            对拼消耗,最差所有人都联合起来对付你
        '''
        major = None
        cnt = 0
        for num in nums:
            if cnt == 0:
                major = num
                cnt = 1
            else:
                cnt += 1 if major == num else -1
        return major


if __name__ == '__main__':
    arr = [int(x) for x in input().strip().split(' ')]
    solution = Solution()
    print(solution.majorityElement(arr))

File: leetcode/array/169.cpp
#include <iostream>

class Solution {
 public:
  int majority_element(int arr[], int len) {
    int major = arr[0];
    int count = 0;
    for (int i = 0; i < len; ++i) {
      if (count == 0) {
        major = arr[i];
      }

      if (major == arr[i]) {
        ++count;
      } else {
        --count;
      }
    }
    return major;
  }
};

int main() {
  int a[7] = {1, 1, 1, 2, 2, 1, 2};

  Solution solution = Solution();
  int major = solution.majority_element(a, 7);
  std::cout << "major: " << major << std::endl;
}